### ISSUE TITLE
CUDA 8 - Unified Memory

### ISSUE THREAD

**Issue by deeplearning-ai-research** [NervanaSystems/neon#275]
Hi,

Wondering if NEON can use CUDA 8 and unified memory for larger than GPU memory access
(CPU and GPU memory on same memory address) ?

It would help significantly to train large NNetwork.

Thanks

Reference:
http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#um-simplifying

**Comment by thestew42** [comment]
Hi, 

It is possible that neon could take advantage of this feature, but it would likely require a lot of work. We use pycuda to access the CUDA API. While pycuda does appear to have support for managed memory, this support is experimental (https://documen.tician.de/pycuda/driver.html#managed-memory) and we have not done any work to support this in neon.

Managed memory has been supported since CUDA 6.0 and is supported by pycuda, but the new page faulting mechanism on Pascal is required to support over-subscribing device memory as you are referring to. This is explained here (https://devblogs.nvidia.com/parallelforall/cuda-8-features-revealed/). It is not clear to me whether this is only a P100 feature or if other Pascal GPUs support it (1080, 1070, etc). Pre-Pascal GPUs (980, Titan X, etc) would not support this feature, even with CUDA 8 installed.

The bottom line is that, although this is certainly possible, it would require several things:

- pycuda support for managed/virtual memory
- changes to neon GPUTensor/NervanaGPU backend for managed memory support
- running on Pascal GPUs

The majority of our GPUs are still Maxwell, so I don't think this is a high priority for us at the moment, but we would welcome contributions from to community to support features such as this. One thing to note is that our upcoming graph backend may affect how this feature is implemented (and may even make it easier).

Thanks,
-Stewart
